### DISCUSSION TITLE
Py5Graphics transparency on save different from Processing Python?

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#287] [Q&A] [answered]
I have tried this, and I get the transparent background I expected:
```python
def setup():
    size(50, 50)
    sf = 10
    f = createGraphics(width * sf, height * sf)
    beginRecord(f) 
    f.scale(sf)
    circle(6, 6, 10)
    endRecord()
    f.save("out.png")
```
![out](https://user-images.githubusercontent.com/3694604/233469703-a387ac4b-0758-479e-b76c-3bd50da11089.png)

But in py5 I get a black background. Even if I try `f.clear()`.

```python
def setup():
    size(50, 50)
    sf = 10
    f = create_graphics(width * sf, height * sf)
    begin_record(f)
    # f.clear()  # doesn't help
    f.scale(sf)
    circle(6, 6, 10)
    end_record()
    f.save("out_py5.png")
```
![out](https://user-images.githubusercontent.com/3694604/233469447-4ab8a6eb-8e54-4f43-9990-cf413d608670.png)

Is this a known issue or is it by design? Am I doing something wrong?

**Comment by Maltiec** [comment, marked as answer]
you should set parameter drop_alpha=False:

f.save("out_py5.png", drop_alpha=False)

by default py5 drops the alpha channel for compatibility with file formats like jpg

**Reply by hx2A** [reply]
> you should set parameter drop_alpha=False:
> 
> f.save("out_py5.png", drop_alpha=False)
> 
> by default py5 drops the alpha channel for compatibility with file formats like jpg

Excellent answer, @Maltiec !

py5 uses the Python library Pillow to save images. Pillow will throw an error if you try to save an image with an alpha channel to a format that does not support it. However, it might be possible to improve this. What if py5 would check the alpha channel before dropping it? If the alpha channel is nothing but 255 values, it adds no value. If it contains other values, it can not drop it, attempt to save with Pillow, and hope that the user picked a file format that supports alpha values.

**Reply by villares** [reply]
Great! Now I understood. Thanks @Maltiec 

@hx2a,  instead of looking at the alpha contents, why not do like Processing and look at the output formats only and drop for JPG and formats that don't have alpha?

Could we do something like this perhaps?
```python
if drop_alpha is None:
    try:
        output_buffer.save(output_file_path, drop_alpha=False)
    except OSError:
        output_buffer.save(output_file_path, drop_alpha=True)
else:
    output_buffer.save(output_file_path, drop_alpha=drop_alpha)

```

**Reply by hx2A** [reply]
> @hx2A, instead of looking at the alpha contents, why not do like Processing and look at the output formats only and drop for JPG and formats that don't have alpha?

Hmmm, that could work, but [Pillow supports many image formats](https://pillow.readthedocs.io/en/stable/handbook/image-file-formats.html), many of which I am not familiar with. Plus, their documentation contains gems like this:

> Pillow reads and writes WebP files. The specifics of Pillow’s capabilities with this format are currently undocumented.

It would be headache to try to figure out which support alpha channels and which don't.

> Could we do something like this perhaps?
> 
> ```python
> if drop_alpha is None:
>     try:
>         output_buffer.save(output_file_path, drop_alpha=False)
>     except OSError:
>         output_buffer.save(output_file_path, drop_alpha=True)
> else:
>     output_buffer.save(output_file_path, drop_alpha=drop_alpha)
> ```

Something like that could work. I'd probably emit a Python warning if an exception is thrown so that the user knows to set drop_alpha to `True`. If someone is saving many frames, they would not want an exception to be thrown and caught every single time (for performance reasons).

**Reply by villares** [reply]
Today the  `drop_alpha` default tripped me again :( 
I had an image buffer showing correctly on screen after `.mask(mask)`, and when saving I got the original image without the masking... it took me a while to figure out it was.... this!

So, one more idea... maybe simpler and without the performance penalty of exception handling:
Make the default `drop_alpha=False` for all formats except 'jpg'/'jpeg' which is the most popular format without alpha.
Then if the user saves to any other alpha-less format they get the error message, and explanation about adding `drop_alpha=True`...

Is this too self-serving? In my defense, the current behavior departs a bit from Processing and could trip other people coming from Processing. I think what bothers me is that it kind of fails in a silent way, giving little clue of what is going on. So maybe to throw an error message on the less common alpha-less formats would be better (did I say maybe?).

**Reply by hx2A** [reply]
I like the idea of improving this but making the default parameter dependent on the image format seems risky. Also Pillow and therefore py5 can save images in a lot of other formats besides jpg and png. and it would be difficult to identify all of the formats that support an alpha channel. And making jpg different from other formats that also don't have alpha channels seems like it would be confusing also.

I like the idea I suggested before:

```python
if drop_alpha is None:
    try:
        output_buffer.save(output_file_path, drop_alpha=False)
    except OSError:
        output_buffer.save(output_file_path, drop_alpha=True)
else:
    output_buffer.save(output_file_path, drop_alpha=drop_alpha)
```

And perhaps this could be improved. What if it remembered which file extensions result in a OSError so it could avoid exception handling the second time it is called? I think this would be possible.

**Comment by villares** [comment]
Cheers @hx2A!

I was trying to show PNG export to someone today, and this didn't work as I expected:

```python
extra_canvas.save('out.png', drop_alhpa=False)
```

I was going nuts, then I saw my mistake. Can you see it?  
I messed the keyword argument name, but the thing is, it failed silently.

Is this a consequence of passing the **kwargs to a delegated Pillow save mechanism? Should we open an issue?

**Reply by hx2A** [reply]
The kwarg `drop_alhpa` is misspelled. The `drop_alpha` kwarg is a py5 kwarg, but if you misspell it, it will be passed along to Pillow instead. If Pillow does not raise an exception when passed unrecognized kwargs, it will be ignored. It is a design decision on their end to either raise an exception or ignore it. I would prefer an exception, and I think that's what matplotlib does? Not sure if I remember correctly. Anyhow, I would check their repo's issues to see if anyone brought this up already, and if not, open an issue to see what they think of it.
